File: src/extractors/be-ast.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# Ruby symbol extractor. Batch protocol: one JSON object per stdin line
#   {"id": "<blob sha>", "path": "app/models/x.rb", "b64": "<base64 source>"}
# emits one JSON line per input, in order:
#   {"id":..., "ok":true, "parser":"ast"|"regex", "symbols":[...], "calls":[...], "includes":{owner:[...]}}
#   {"id":..., "ok":false, "error":"..."}
# Uses RubyVM::AbstractSyntaxTree (stdlib, Ruby 2.6+). No gems. Files the running Ruby cannot parse
# (newer syntax) fall back to a line scanner that still recovers class/module/def declarations.
require 'json'
require 'base64'

Node = RubyVM::AbstractSyntaxTree::Node

def const_name(n)
  return nil unless n.is_a?(Node)
  case n.type
  when :CONST  then n.children[0].to_s
  when :COLON3 then n.children[0].to_s
  when :COLON2
    base = const_name(n.children[0])
    base ? "#{base}::#{n.children[1]}" : n.children[1].to_s
  else nil
  end
end

def join_ns(ns, name)
  return name if ns.empty?
  "#{ns.join('::')}::#{name}"
end

class Extractor
  def initialize
    @symbols = []; @calls = []; @includes = Hash.new { |h, k| h[k] = [] }
  end

  def run(root)
    walk(root, [], false, nil)
    { symbols: @symbols, calls: @calls, includes: @includes }
  end

  private

  def sym(kind, name, n, owner = nil, extra = {})
    @symbols << { kind: kind, name: name, owner: owner, line: n.first_lineno, end_line: n.last_lineno }.merge(extra)
  end

  def walk(n, ns, singleton, cur)
    return unless n.is_a?(Node)
    case n.type
    when :CLASS
      name = const_name(n.children[0]) || '?'
      full = join_ns(ns, name)
      sym('class', full, n, nil, superclass: const_name(n.children[1]))
      walk(n.children[2], ns + [name], false, nil)
    when :MODULE
      name = const_name(n.children[0]) || '?'
      full = join_ns(ns, name)
      sym('module', full, n)
      walk(n.children[1], ns + [name], false, nil)
    when :SCLASS
      recv = n.children[0]
      walk(n.children[1], ns, true, nil) if recv.is_a?(Node) && recv.type == :SELF
    when :DEFN
      owner = ns.join('::'); mname = n.children[0].to_s
      full = owner.empty? ? mname : "#{owner}#{singleton ? '.' : '#'}#{mname}"
      sym(singleton ? 'class_method' : 'method', full, n, owner)
      walk(n.children[1], ns, singleton, full)
    when :DEFS
      recv = n.children[0]; mname = n.children[1].to_s; owner = ns.join('::')
      target = (recv.is_a?(Node) && recv.type == :SELF) ? owner : (const_name(recv) || owner)
      full = target.empty? ? mname : "#{target}.#{mname}"
      sym('class_method', full, n, target)
      walk(n.children[2], ns, true, full)
    when :FCALL, :VCALL
      mname = n.children[0].to_s
      if cur.nil? && %w[include extend prepend].include?(mname) && !ns.empty?
        args = n.children[1]
        if args.is_a?(Node)
          args.children.each { |a| c = const_name(a); @includes[ns.join('::')] << "#{mname}:#{c}" if c }
        end
      elsif cur
        @calls << { from: cur, recv: nil, meth: mname, line: n.first_lineno }
      end
      n.children.each { |c| walk(c, ns, singleton, cur) }
    when :CALL, :QCALL, :OPCALL
      if cur
        recv = const_name(n.children[0])
        @calls << { from: cur, recv: recv, meth: n.children[1].to_s, line: n.first_lineno } if recv || n.type != :OPCALL
      end
      n.children.each { |c| walk(c, ns, singleton, cur) }
    else
      n.children.each { |c| walk(c, ns, singleton, cur) }
    end
  end
end

# Line scanner for files the running Ruby cannot parse (Ruby 3 syntax on Ruby 2.6, etc.).
# Nesting by indentation. Recovers declarations only, no calls. Marked as such.
def regex_extract(src)
  symbols = []; stack = [] # [indent, name]  -- popped ONLY by an `end` at the same indent,
  in_sclass = nil          # so heredoc bodies and other column-0 text cannot close a class.
  src.each_line.with_index(1) do |raw, ln|
    line = raw.rstrip
    next if line.strip.empty? || line.strip.start_with?('#')
    indent = line[/\A */].size
    if line =~ /\A\s*end\b/
      in_sclass = nil if in_sclass && indent <= in_sclass
      stack.pop if stack.any? && stack.last[0] == indent
      next
    end
    if (m = line.match(/\A\s*(class|module)\s+([A-Z][\w:]*)(?:\s*<\s*([\w:]+))?/))
      kind, name, sup = m[1], m[2], m[3]
      full = stack.empty? ? name : "#{stack.map { |s| s[1] }.join('::')}::#{name}"
      symbols << { kind: kind, name: full, owner: nil, line: ln, end_line: nil, superclass: sup }
      stack << [indent, name] unless line =~ /;\s*end\s*\z/   # one-liner `class X < Y; end`
    elsif line =~ /\A\s*class\s*<<\s*self\b/
      in_sclass = indent
    elsif (m = line.match(/\A\s*(?:(?:private|protected|public|private_class_method|memoize|memoized|sig)\s+)*def\s+(self\.)?([\w?!=<>+\-*\/%\[\]~^&|]+)/))
      owner = stack.map { |s| s[1] }.join('::')
      singleton = !m[1].nil? || (in_sclass && indent > in_sclass)
      full = owner.empty? ? m[2] : "#{owner}#{singleton ? '.' : '#'}#{m[2]}"
      symbols << { kind: singleton ? 'class_method' : 'method', name: full, owner: owner, line: ln, end_line: nil }
    end
  end
  { symbols: symbols, calls: [], includes: {} }
end

$stdout.sync = true
$stdin.each_line do |raw|
  raw = raw.strip
  next if raw.empty?
  item = JSON.parse(raw)
  src = Base64.decode64(item['b64']).force_encoding('UTF-8')
  src = src.scrub('?') unless src.valid_encoding?
  begin
    ast = RubyVM::AbstractSyntaxTree.parse(src)
    out = Extractor.new.run(ast)
    puts JSON.generate({ id: item['id'], ok: true, parser: 'ast' }.merge(out))
  rescue SyntaxError => e
    begin
      out = regex_extract(src)
      puts JSON.generate({ id: item['id'], ok: true, parser: 'regex', syntax_error: e.message.lines.first.to_s.strip[0, 200] }.merge(out))
    rescue StandardError => e2
      puts JSON.generate({ id: item['id'], ok: false, error: "#{e.class}: #{e.message.lines.first.to_s.strip[0, 200]}; fallback: #{e2.message[0, 100]}" })
    end
  rescue StandardError => e
    puts JSON.generate({ id: item['id'], ok: false, error: "#{e.class}: #{e.message[0, 200]}" })
  end
end

File: src/extractors/be-ast.mjs
// PER-BLOB extractor: Ruby symbols (classes, modules, methods) via the stdlib AST.
// Replaces the one-off Ruby AST dump the graph was bootstrapped from, so a re-index of any
// backend commit regenerates every SYMBOL node itself. Identities match the import exactly:
//   be:sym:Api::V1::TradeController            (class / module)
//   be:sym:Api::V1::TradeController#quote_details   (instance method)
//   be:sym:CustomConfiguration.cached_all_configs   (class method)
// Batched: one Ruby process parses every changed file of a run (thousands on a cold start,
// a handful on a merge). Files the running Ruby cannot parse fall back to a line scanner and
// are marked HEURISTIC so answers can say so.
import { spawnSync } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

export const NAME = "be-ast";
export const VERSION = "1.0";
const HERE = path.dirname(fileURLToPath(import.meta.url));
const SCRIPT = path.join(HERE, "be-ast.rb");

// What the bootstrap dump covered: app code, lib, migrations, one-off scripts. Not specs, config, vendor.
export function handles(p) {
  return /^(app|lib|db\/migrate|scripts|mcp_tools)\/.*\.rb$/.test(p) && !/^app\/(assets|views)\//.test(p);
}

/** items: [{sha, buf, path}] -> Map(sha -> {facts, ok, error}) */
export function extractBatch(items) {
  const input = items.map(it => JSON.stringify({ id: it.sha, path: it.path, b64: Buffer.from(it.buf).toString("base64") })).join("\n") + "\n";
  const r = spawnSync("ruby", [SCRIPT], { input, encoding: "utf8", maxBuffer: 1024 * 1024 * 1024 });
  if (r.status !== 0 && !r.stdout) throw new Error(`be-ast.rb failed: ${(r.stderr || "").slice(0, 400)}`);
  const out = new Map();
  for (const line of r.stdout.split("\n")) {
    if (!line.trim()) continue;
    let o; try { o = JSON.parse(line); } catch { continue; }
    if (o.ok) out.set(o.id, { ok: true, facts: { schema: 1, parser: o.parser, syntax_error: o.syntax_error, symbols: o.symbols, calls: o.calls, includes: o.includes } });
    else out.set(o.id, { ok: false, error: o.error, facts: { schema: 1, parser: "none", symbols: [], calls: [], includes: {} } });
  }
  for (const it of items) if (!out.has(it.sha)) out.set(it.sha, { ok: false, error: "no output from be-ast.rb", facts: { schema: 1, parser: "none", symbols: [], calls: [], includes: {} } });
  return out;
}

/** Single-blob form of the same thing, for callers that do not batch. */
export function extract(buf, p) {
  const r = extractBatch([{ sha: "one", buf, path: p }]).get("one");
  if (!r.ok) throw new Error(r.error);
  return r.facts;
}

// Rails' `underscore`: "Api::V1::ApprovalWorkflow::TradeController" -> "api/v1/approval_workflow/trade"
const underscore = (klass) => klass.replace(/Controller$/, "").split("::")
  .map(s => s.replace(/([A-Z\d]+)([A-Z][a-z])/g, "$1_$2").replace(/([a-z\d])([A-Z])/g, "$1_$2").toLowerCase()).join("/");

export function resolve(facts, file, ctx) {
  const entities = [], edges = [];
  if (!facts?.symbols) return { entities, edges };
  const heur = facts.parser === "regex";
  const includes = facts.includes || {};
  const seen = new Set();

  for (const s of facts.symbols) {
    if (!s.name || s.name.includes("?::") || seen.has(s.name)) continue;
    // A top-level `def` in a script has no owner and its name collides across files (get_config x4).
    // The bootstrap dump carried none of these; neither do we.
    if ((s.kind === "method" || s.kind === "class_method") && !s.owner) continue;
    seen.add(s.name);
    const fqn = `be:sym:${s.name}`;
    const attrs = { subkind: s.kind, parser: facts.parser };
    if (s.owner) attrs.class = s.owner;
    if (s.superclass) attrs.superclass = s.superclass;
    if ((s.kind === "class" || s.kind === "module") && includes[s.name]?.length) attrs.mixins = includes[s.name];
    entities.push({ fqn, kind: "SYMBOL", name: s.name, path: file.path, blobSha: file.blobSha,
                    startLine: s.line, endLine: s.end_line ?? null, attrs,
                    status: "OBSERVED", extractor: `${NAME}@${VERSION}`,
                    confidence: heur ? 0.8 : 1.0, resolution: heur ? "HEURISTIC" : "EXACT" });

    if ((s.kind === "method" || s.kind === "class_method") && s.owner) {
      // class --DECLARES--> method. Same hash as the bootstrap import, so nothing duplicates on the pinned commit.
      edges.push({ kind: "DECLARES", srcFqn: `be:sym:${s.owner}`, dstFqn: fqn,
                   siteHash: ctx.siteHash(`${s.owner}|${fqn}`), startLine: null,
                   confidence: 1.0, resolution: heur ? "HEURISTIC" : "EXACT" });
      // HANDLER (from routes) --DECLARES--> the controller action that implements it.
      // Only lands if the routes extractor produced that handler (edges are filtered on both ends at load).
      if (s.kind === "method" && /Controller$/.test(s.owner)) {
        const handlerFqn = `be:${underscore(s.owner)}#${s.name.split("#").pop()}`;
        edges.push({ kind: "DECLARES", srcFqn: handlerFqn, dstFqn: fqn,
                     siteHash: ctx.siteHash(`${handlerFqn}|${fqn}`), startLine: null,
                     confidence: 0.95, resolution: "EXACT" });
      }
    }
  }

  // Static call edges are OFF by default: they are name-based guesses, and the graph's runtime
  // (observed) CALLS edges are the trustworthy ones. CKG_STATIC_CALLS=1 adds `Const.method` calls as HEURISTIC.
  if (process.env.CKG_STATIC_CALLS === "1") {
    for (const c of facts.calls || []) {
      if (!c.recv || !c.meth || !c.from) continue;
      edges.push({ kind: "CALLS", srcFqn: `be:sym:${c.from}`, dstFqn: `be:sym:${c.recv}.${c.meth}`,
                   siteHash: ctx.siteHash(`${c.from}|${c.recv}.${c.meth}|${c.line}`), startLine: c.line,
                   confidence: 0.7, resolution: "HEURISTIC" });
    }
  }
  return { entities, edges };
}
